[DCP] DCP logger

LucasLLC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121352

Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

LucasLLC committed:
[DCP] RFC, first pass at a DCP logger

This diff represents a RFC on the overall approach of improving DCP observability and debug ability internally. Any lingering TODO's, tests, or general code cleanliness will be taken care of before we merge.

A large motivator for this work is unification with model store components. To that extent, please note any existing model store logs should not be changing. In the "Unified Checkpointer", existing model store logs should be be migrated either to one of the plugins passed to DCP API's (StorageReader/Writer, Save/LoadPlanner), or to the trainer level.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] RFC, first pass at a DCP logger"


This diff represents a RFC on the overall approach of improving DCP observability and debug ability internally. Any lingering TODO's, tests, or general code cleanliness will be taken care of before we merge.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] RFC, first pass at a DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wconstab commented on torch/distributed/checkpoint/logger.py:
```diff
@@ -0,0 +1,92 @@
+            # log start event
+            msg_dict["event"] = "start"
+            t0 = time.time_ns()
+            msg_dict["time"] = t0
```
is it possible for t0 to ever be observed as msg_dict["time"]?  it looks like it would always be replaced by exception or completion time.

LucasLLC replied:
So `msg_dict` is logged as is in the `_dcp_logger.debug` call below. At that point it should have the correct value. 

The wrapper should create three distinct logs -- start, end, & exception

wconstab reviewed (commented):
I think the big question i have for this is about what c10d_logger does under the hood and whether we want these logs as 'logs' or we want them as 'metrics' that go into some kind of DB.

for the latter, we might want to integrate with otel later instead of doing scuba now.  But just wondering.

I think the mechanics of using the decorator to log the API calls looks fine.

LucasLLC commented:
> I think the big question i have for this is about what c10d_logger does under the hood and whether we want these logs as 'logs' or we want them as 'metrics' that go into some kind of DB.
> 
> for the latter, we might want to integrate with otel later instead of doing scuba now. But just wondering.
> 
> I think the mechanics of using the decorator to log the API calls looks fine.

Hm, I think they generally serve both purposes. We need metrics to be collected so we can track telemetry for how long some methods are taking in distributed environments. We also need logs we can use to debug when things go wrong.

Otel does look interesting, I agree we might want to integrate with otel later, but scuba unblocks us now.

LucasLLC commented:
> c10d_logger

> i am trying to understand how it works. does c10d_logger under the hood log everything to both scuba and console? or is there some flag that decides which one?

In internal environments, due to a build trick in the TARGETS file,  c10d logs both to scuba and signpost. In OSS, the defualt handler is a NullHandler, but I believe are still logged to console via the regular stream handler (I'll confirm). Let me know if that clears things up

wz337 reviewed (approved):
Great! Have discussed the RFC together and this aligns with it. LGTM!

fegin commented on torch/distributed/checkpoint/logger.py:
```diff
@@ -0,0 +1,92 @@
+_P = ParamSpec("_P")
+
+
+def _parse_dcp_method_args(*args, **kwargs) -> Tuple[Tuple[Any], Dict[str, Any]]:
```
Since this is eventually used in a functional wrapper, you can use `inspect` to inspect the function arguments and keyword arguments. This is just a suggestion and you don't have to change it in this PR. And this may be useful if we want to use the same logger wrapper for distributed state_dict.

LucasLLC replied:
Good idea, thanks!

fegin reviewed (approved):
Are we going to use this for distributed state_dict?

LucasLLC commented:
> Are we going to use this for distributed state_dict?

I think that's a solid idea. We'll have to see how that might change the implementation

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC committed:
Update on "[DCP] DCP logger"


Adds additional logging for improved observability in DCP.

Differential Revision: [D54512626](https://our.internmc.facebook.com/intern/diff/D54512626/)


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

LucasLLC commented:
@pytorchbot merge

Skylion007 commented on torch/distributed/checkpoint/logging_handlers.py:
```diff
@@ -0,0 +1,16 @@
+from torch.distributed.logging_handlers import _log_handlers
+
+
+print("in oss file")
```
Was this print statement suppose be merged @LucasLLC ?
